sklearn/tree/_tree.pyx

```diff
@@ -777,23 +805,59 @@ cdef class Tree:
         cdef Node* node = NULL
         cdef SIZE_t i = 0
 
-        with nogil:
-            for i in range(n_samples):
-                node = self.nodes
-                # While node not a leaf
-                while node.left_child != _TREE_LEAF:
-                    # ... and node.right_child != _TREE_LEAF:
-                    if X_ptr[X_sample_stride * i +
-                             X_fx_stride * node.feature] <= node.threshold:
-                        node = &self.nodes[node.left_child]
-                    else:
-                        node = &self.nodes[node.right_child]
-
-                out_ptr[i] = <SIZE_t>(node - self.nodes)  # node offset
+        cdef BOOL_t* missing_mask_ptr
+        cdef SIZE_t missing_mask_fx_stride
+        cdef UINT32_t* random_state
+        cdef int temp
+
+        if not self.allow_missing:
+            with nogil:
+                for i in range(n_samples):
+                    node = self.nodes
+                    # While node not a leaf
+                    while node.left_child != _TREE_LEAF:
+                        # ... and node.right_child != _TREE_LEAF:
+                        if X_ptr[X_sample_stride * i +
+                                 X_fx_stride * node.feature] <= node.threshold:
+                            node = &self.nodes[node.left_child]
+                        else:
+                            node = &self.nodes[node.right_child]
+
+                    out_ptr[i] = <SIZE_t>(node - self.nodes)  # node offset
+        else:
+            # Extract the missing mask
+            missing_mask_ptr = <BOOL_t*> missing_mask.data
+            missing_mask_fx_stride = <SIZE_t> missing_mask.strides[1] / <SIZE_t> missing_mask.itemsize
+            random_state = &self.rand_r_state
+            with nogil:
+                for i in range(n_samples):
+                    node = self.nodes
+                    # While node not a leaf
+                    while node.left_child != _TREE_LEAF:
+                        # ... and node.right_child != _TREE_LEAF:
+                        if missing_mask_ptr[i + missing_mask_fx_stride * node.feature] == 1:
+                            if node.missing_direction == MISSING_DIR_UNDEF:
```

Comment:
This is when the training data has no missing value but the testing  data has. (This could happen for the  whole tree or for some nodes alone depending on the distribution of the missing values)